modules/base/tool_test.go

```diff
@@ -129,6 +129,38 @@ func TestEllipsisString(t *testing.T) {
 	assert.Equal(t, "测试文本一二三四", EllipsisString("测试文本一二三四", 10))
 }
 
+func TestEllipsisStringWholeWord(t *testing.T) {
+	assert.Equal(t, "…", EllipsisStringWholeWord("foobar", 0))
+	assert.Equal(t, "f…", EllipsisStringWholeWord("foobar", 1))
+	assert.Equal(t, "fo…", EllipsisStringWholeWord("foobar", 2))
+	assert.Equal(t, "foo…", EllipsisStringWholeWord("foobar", 3))
+	assert.Equal(t, "foob…", EllipsisStringWholeWord("foobar", 4))
+	assert.Equal(t, "fooba…", EllipsisStringWholeWord("foobar", 5))
+	assert.Equal(t, "foobar", EllipsisStringWholeWord("foobar", 6))
+	assert.Equal(t, "foobar", EllipsisStringWholeWord("foobar", 10))
+
+	assert.Equal(t, "…", EllipsisStringWholeWord("foo bar", 0))
+	assert.Equal(t, "f…", EllipsisStringWholeWord("foo bar", 1))
+	assert.Equal(t, "fo…", EllipsisStringWholeWord("foo bar", 2))
+	assert.Equal(t, "foo…", EllipsisStringWholeWord("foo bar", 3))
+	assert.Equal(t, "foo…", EllipsisStringWholeWord("foo bar", 4))
+	assert.Equal(t, "foo…", EllipsisStringWholeWord("foo bar", 5))
+	assert.Equal(t, "foo…", EllipsisStringWholeWord("foo bar", 6))
+	assert.Equal(t, "foo bar", EllipsisStringWholeWord("foo bar", 7))
+	assert.Equal(t, "foo bar", EllipsisStringWholeWord("foo bar", 10))
+
+	assert.Equal(t, "foo bar…", EllipsisStringWholeWord("foo bar foo", 7))
+	assert.Equal(t, "foo bar…", EllipsisStringWholeWord("foo bar foo", 8))
+	assert.Equal(t, "foo bar…", EllipsisStringWholeWord("foo bar foo", 9))
+	assert.Equal(t, "foo bar…", EllipsisStringWholeWord("foo bar foo", 10))
+	assert.Equal(t, "foo bar foo", EllipsisStringWholeWord("foo bar foo", 11))
+
+	assert.Equal(t, "测试文本…", EllipsisStringWholeWord("测试文本一二三四", 4))
```

Comment:
The truncation should be done at "byte" level, but not "rune" level.

The problem of "rune" level is that CJK chars will be too much longer than expected.

```
assert.Equal(t, "foo…", EllipsisStringWholeWord("foo bar", 4))
assert.Equal(t, "测试文本…", EllipsisStringWholeWord("测试文本一二三四", 4))
```

See the difference. That's why the old function is "SplitStringAt**Byte**N"

Reply:
Actually there is usually no "database restriction", because `VCHAR(10)` could store 10 CJK runes or emoji runs (say, 30 - 40bytes).

I think the real problem is that CJK chars render much longer than ASCII chars with the same length on the UI ......... for example, assume one ASCII char's width is about _W_.

* 9 CJK runes: about _W_ * 18
* 9 bytes CJK chars: _W_ * 3 * 2 = _W_ * 6
* 9 bytes ASCII chars: _W_ * 9

So "9 CJK runes" exceeds the width limitation too much on the UI, while usually "9 bytes CJK chars" won't exceed the expected "9 bytes ASCII" width.

Reply:
->  Refactor "string truncate" #32984